### PR TITLE
[proxy] Add support of `DD_` prefixed env vars

### PR DESCRIPTION
**Pull request by olivielpeau**
Update:
* using env var naming that's closer to Agent 6 convention
* use empty env var values when they're present
* use space-separated list for `DD_PROXY_NO_PROXY`

### What does this PR do?

Add support of `DD_PROXY_HTTP`, `DD_PROXY_HTTPS` and `DD_PROXY_NO_PROXY` env vars.

They take precedence respectively over the `HTTP_PROXY`, `HTTPS_PROXY` and `NO_PROXY` env vars, which themselves take precedence over the `proxy` settings defined directly in `datadog.yaml`.

So overall: `DD_`-env > standard env > `datadog.yaml`

Also, the proxy env vars with empty values are now used instead of being ignored. This is a minor breaking change, but should allow users to meaningfully set an env var to an empty value when they actually want the value used by the Agent to be empty.

### Motivation

The `HTTP_PROXY`, `HTTPS_PROXY` and `NO_PROXY` env vars are "standards" that are used automatically by many libs and tools (for example the `requests` python lib).

When users use `HTTP_PROXY`, `HTTPS_PROXY` and `NO_PROXY` to set up the Agent proxy, many integrations (e.g. `kubelet`) that use `requests` without passing explicit `proxies` settings will also end up using these proxy settings. That's a breaking and undesirable change because, before 6.3.0, these integrations didn't use proxies when the Agent was set up to use a proxy, and these integrations are generally not meant to be used with proxies (they hit internal domains/IPs).

So let's introduce these new DD-specific env vars that have the exact same effect as setting `proxy` in `datadog.yaml`: they will only be used by the Agent and the integrations that want to use the Agent's proxy settings, but will not make the other integrations use them.

We keep support of the `HTTP_PROXY`, `HTTPS_PROXY` and `NO_PROXY` env vars to be consistent with the "standard" (i.e. what any other tool would do) and with what the Agent 5 does.

### Additional Notes

~**Open questions**:~ _Questions answered, see update at top of PR description._

~1. Should we rather use `DD_PROXY_HTTP`, `DD_PROXY_HTTPS` and `DD_PROXY_NO_PROXY` as the new env var keys? It would be consistent with the naming of the other `DD_` env vars (which is based on the `datadog.yaml` name). I'd be in favor of using them for consistency.~
~2. When an env var is present and set to `""`, should we use its empty value or should we discard it? Currently we discard it (and continue reading the option with less precedence), but it may not be desirable as a user may want to explicitly set `DD_HTTP_PROXY` to `""` while `HTTP_PROXY` is set to a non-empty value in order to override the latter.~

TODO:

* [x] add release note
* [x] update `datadog-agent` docs:
  - https://github.com/DataDog/datadog-agent/blob/master/Dockerfiles/agent/README.md
  - https://github.com/DataDog/datadog-agent/blob/master/docs/agent/changes.md#L98
  - https://github.com/DataDog/datadog-agent/blob/6.3.3/pkg/config/config_template.yaml#L14
* [x] Make sure `process-agent` and `trace-agent` aren't impacted/are consistent.

### CODE REVIEW

File: pkg/config/config.go
Comment (hush-hush) [inline pkg/config/config.go]: if `DD_PROXY_HTTP` is set (ie: manual action from the user): I guess we should use it. This would mean not using any proxy.

We could write an info message in that case. But it might be less confusing then using `HTTP_PROXY` when you did set `DD_PROXY_HTTP`.

We could use `os.LookupEnv` for this.

File: pkg/config/config.go
Comment (hush-hush) [inline pkg/config/config.go]: I would indeed be in favor of `DD_PROXY_HTTP` to be consistent with the rest of the configuration. `HTTP_PROXY` is an exception because it's kind of a standard on linux.

**Review by hush-hush — APPROVED** [review]
I'm OK with this solution as long as we clearly explain how to set a proxy in our documentation.

Also we could/should add an info log line at the start of the agent printing the final proxy settings.

We might want to tell people to use `DD_*` variables to set the proxy. And in addition we support standard `HTTP(S)_PROXY` variables, with a warning about the impact on the checks.

**Review by hush-hush — CHANGES_REQUESTED** [review]
`DD_HTTP_PROXY` -> `DD_PROXY_HTTP`

File: pkg/config/config.go
Comment (hkaj) [inline pkg/config/config.go]: don't we want to be case insensitive on this `key` as well?

Comment (olivielpeau) [reply pkg/config/config.go]: `viper` is case-sensitive on the env variables. Since this env var is meant to be pulled like any other env var pulled by `viper` I think it's better to be case-sensitive for consistency.

I can add a comment to clarify this if you think it's worth it

File: pkg/config/config.go
Comment (olivielpeau) [reply pkg/config/config.go]: makes sense, I'll implement this

File: pkg/config/config_test.go
Comment (olivielpeau) [inline pkg/config/config_test.go]: (btw, I modified this test because I think it wasn't testing something interesting)

File: pkg/config/config_test.go
Comment (hush-hush) [inline pkg/config/config_test.go]: Could we add a test on `NO_PROXY` since the split is different from `DD_NO_PROXY` (space vs coma).

**Review by hush-hush — CHANGES_REQUESTED** [review]
I'm ok with the implementation :+1: 

But I think we should update those docs:

- https://github.com/DataDog/datadog-agent/blob/master/docs/agent/changes.md#proxies
- https://github.com/DataDog/datadog-agent/tree/master/Dockerfiles/agent#proxies (seems like a duplicate, maybe we should like the first doc in the  README of the Dockerfile ?)

File: pkg/config/config_test.go
Comment (olivielpeau) [reply pkg/config/config_test.go]: other tests cover this, but I can create a specific test just for this yes

Comment (olivielpeau) [reply pkg/config/config_test.go]: actually I think this is covered well-enough by the `TestLoadProxyStdEnvOnly` and `TestLoadProxyDDSpecificEnvOnly` tests, I added a few comments there to make it more obvious (02be916)

**Comment by olivielpeau** [comment]
The PR is ready:

* `trace-agent` does _not_ support the `DD_` env vars at the moment. I've documented this in the Agent Dockerfile's README.
* `process-agent` should support the `DD_` env vars once we merge the PR and they update the dependency on `datadog-agent`. Waiting on confirmation from them.

**Review by hush-hush — DISMISSED** [review]

File: docs/agent/changes.md
Comment (zippolyte) [inline docs/agent/changes.md]: Do we want to specify the precedence of `DD_PROXY_*` over the standard variables ?

**Review by hkaj — DISMISSED** [review]
`Dockerfiles` update looks good 👍

File: docs/agent/changes.md
Comment (olivielpeau) [reply docs/agent/changes.md]: I wanted to avoid mentioning them, but you're right it's something we should document. I've updated the paragraph above, let me know what you think

**Review by zippolyte — DISMISSED** [review]
Sounds good to me

**Comment by olivielpeau** [comment]
just realized we can't log anything directly from `loadProxyFromEnv` since when it's called the logger isn't initialized yet. These log entries are pretty useful so I'll try to implement a workaround

**Review by hush-hush — APPROVED** [review]

**Comment by olivielpeau** [comment]
@hush-hush will address the log issue in `loadProxyFromEnv` in a separate PR, I'll merge this first :)

**Pull request merged**
